manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+guarantees in more cases.
+
+\paragraph{Performance} A standard approach to get a tail-recursive
+version of "List.map" is to use an accumulator to collect output
+elements, and reverse it at the end of the traversal.
+
+\begin{verbatim}
+let rec map f li = map_aux f [] li
+and map_aux f acc li =
+  match li with
+  | [] -> List.rev acc
+  | x :: xs ->
+    let y = f x in
+    map_aux f (y :: acc) xs
+\end{verbatim}
+
+This version is tail-recursive, but it is measurably slower than the
+simple, non-tail-recursive version. In contrast, the tail-mod-cons
+transformation provides an implementation that has comparable
+performance to the original version, even on small inputs.
+
+\paragraph{Evaluation order} Beware that the tail-modulo-cons
+transformation has an effect on evaluation order: the constructor
+argument that is transformed into tail-position will always be
+evaluated last. Consider the following example:
+
+\begin{verbatim}
+type 'a two_headed_list =
+  | Nil
+  | Consnoc of 'a * 'a two_headed_list * 'a
+
+let[@tail_mod_cons] rec map f = function
+  | Nil -> Nil
+  | Consnoc (front, body, rear) ->
+    Consnoc (f front, map f body, f rear)
+\end{verbatim}
+
+Due to the "[\@tail_mod_cons]" transformation, the calls to "f front"
+and "f rear" will be evaluated before "map f body". In particular,
+this is likely to be different from the evaluation order of the
+un-annotated version. (The evaluation order of constructor arguments
+is unspecified in OCaml, but many implementations typically use
+left-to-right or right-to-left.)
+
+This effect on evaluation order is one of the reasons why the
+tail-modulo-cons transformation has to be explicitly requested by the
+user, instead of being applied as an automatic optimization to some
+functions.
+
+\paragraph{Why tail-modulo-cons?} Other program transformations, in
+particular a transformation to continuation-passing style (CPS), can
+make all functions tail-recursive, instead of targeting only a small
+fragment. Some reasons to provide builtin support for the less-general
+tail-mod-cons are as follows:
+\begin{itemize}
+\item The tail-mod-cons transformation preserves the performance of
+  the original, non-tail-recursive version, while
+  a continuation-passing-style transformation incurs a measurable
+  constant-factor overhead.
+\item The tail-mod-cons transformation cannot be expressed as
+  a source-to-source transformation of OCaml programs, as it relies on
+  mutable state in type-unsafe ways. In contrast,
+  continuation-passing-style versions can be written by hand, possibly
+  using a convenient monadic notation.
+\end{itemize}
+
+\subsection{ss:disambiguation}{Disambiguation}
+
+It may happen that several arguments of a constructor are recursive
+calls to a tail-modulo-cons function. The transformation can only turn
+one of these calls into a tail call. It will not make a choice
+implicitly, but ask the user to disambiguate explicitly.
```

Comment:
```suggestion
implicitly, but ask the user to provide an explicit disambiguation.
```
? (though I like that you sort to not split your infinitives! 🙂)

Reply:
```tex
Instead of making an implicit choice, the compiler will ask the user to provide an explicit disambiguation.
```
?

Reply:
Thanks both for the suggestions. I went for

> The compiler will not make an
implicit choice, but ask the user to provide an explicit
disambiguation.

(I liked the symmetry between "implicitly" and "explicitly" we had before, and in this version inspired from @ksromanov we keep the symmetry between "implicit" and "explicit".)